### DISCUSSION TITLE
Entity class name is being used in executed query instead of table name in `findOne`, and it keeps failing.

### DISCUSSION THREAD

**Discussion by premdasvm** [mikro-orm/mikro-orm#4083] [Q&A] [answered]
i've a custom pivot table entity as follows,

```ts
@Entity({ tableName: "user_shift" })
export class UserShift extends BaseEntity {
    @ManyToOne({ primary: true })
    user: User;

    @ManyToOne({ primary: true })
    shift: Shift;

    @Property({ type: "json" })
    coordinates?: IGeolocation;
}
```
`user.entity.ts`
```ts
 @ManyToMany({
   entity: () => Shift,
   pivotEntity: () => UserShift,
})
shifts = new Collection<UserShift>(this);
```
`shift.entity.ts`

```ts
@ManyToMany({ entity: () => User, mappedBy: u => u.shifts })
users = new Collection<User>(this);
```

these are my entities, Now the issue i'm facing is when ever i call this function

```ts
const user = await this.userRepo.findOne(id, { populate: ["shifts"] });
```
I get this error(check the image). but if i don't populate "shifts" it works fine. any idea what i'm supposed to do?

![Screenshot 2023-03-01 at 4 47 36 PM](https://user-images.githubusercontent.com/14838550/222124599-72abd8b7-cde2-4ac3-be8f-50165e16ec3a.png)

**Comment by B4nan** [comment, marked as answer]
What version are you on?

**Reply by premdasvm** [reply]
5.6.11

**Reply by premdasvm** [reply]
Actually it looks like the query i'm trying to execute is being executed successfully. But i've no idea where the query which is failing is being called from. 🤔

`select "s0".* from "shift" as "s0" where "UserShift"."user" in (1)`

**Reply by premdasvm** [reply]
@B4nan setting `disableIdentityMap: true` seems to fix this issue. any idea why?

**Reply by B4nan** [reply]
I did a quick test and it works fine in the ORM test suite, can you prepare a complete reproduction?

**Reply by premdasvm** [reply]
I'll try to make reproduction.

Is the way i created my entities correct?

because when i don't pass in `{ populate: ["shifts"] }` i still get the joined data on my query and returning data. But if i start specifying fields, then error occurs.

```ts
const user = await this.userRepo.findOne(id, {
 fields: ["shifts.shift.name"],
});
```

 
![Screenshot 2023-03-01 at 5 34 39 PM](https://user-images.githubusercontent.com/14838550/222134659-82f110ab-457c-4969-b2b4-1d221dc560f5.png)

**Reply by B4nan** [reply]
Oh now I see it! You got the relation wrong, the m:n collection contains shift entities, not the pivot ones, and I think that may have let you write wrong queries, as the one you just provided with the `fields` - as if you specify only fields, it is also used as populate hint automatically. 

But that itself should not produce the issue with a wrong query in your opening post, as it is just about type level - but maybe you were oversimplifying? 

```diff
 @ManyToMany({
   entity: () => Shift,
   pivotEntity: () => UserShift,
})
-shifts = new Collection<UserShift>(this);
+shifts = new Collection<Shift>(this);
```

If you want the pivot entities, it is not a m:n relation, those have 1:m - m:1 between each other.

This test defines both m:n and 1:m relations with the pivot entity:

https://github.com/mikro-orm/mikro-orm/blob/master/tests/features/composite-keys/custom-pivot-entity.sqlite.test.ts

**Reply by premdasvm** [reply]
@B4nan Wow, That seems to have fixed my initial issue also.  That wrong query is not being executed now.

But even tho i've specified `fields` the return result is the whole object, 

```ts
const user = await this.userRepo.findOne(id, {
 fields: ["*", "shifts.name", "shifts.startTime"],
});
```

My return object
```json
{
  "id": 1,
  "createdAt": "2023-03-01T11:10:34.000Z",
  "name": "Prem",
  "mobileNumber": "1234567890",
  "phoneCountryCode": "+91",
  "email": "example@gamil.com",
  "isActive": true,
  "role": "ADMIN",
  "shifts": [
    {
      "id": 1,
      "createdAt": null,
      "name": "Morning",
      "startTime": "12:20:00",
      "endTime": null,
      "isActive": null
    }
  ]
}
```
query being executed
```sql
select "u0".*, "s1"."id" as "s1__id", "s1"."name" as "s1__name", "s1"."start_time" as "s1__start_time"
from "user" as "u0"
         left join "user_shift" as "u2" on "u0"."id" = "u2"."user_id"
         left join "shift" as "s1" on "u2"."shift_id" = "s1"."id"
where "u0"."id" = 1
```

any idea on why it's returning the `fields` i didn't ask for as null?

**Reply by B4nan** [reply]
might be some hydration issue, try dropping the `*`, you shouldn't need that

**Reply by premdasvm** [reply]
@B4nan Also in my `user` entity

```ts
@ManyToMany({
 entity: () => Shift,
 pivotEntity: () => UserShift,
 eager: true,
})
shifts = new Collection<Shift>(this);
```

if i remove the `eager : true` then the wrong query is being executed again

```sql
select "s0"."id", "s0"."name", "s0"."start_time", "s0"."end_time" from "shift" as "s0" where "UserShift"."user" in (1) - missing FROM-clause entry for table "UserShift"
```

**Reply by B4nan** [reply]
As I said, provide repro, it works fine in the ORM tests, so it must be about something you are not sharing.

**Reply by premdasvm** [reply]
> might be some hydration issue, try dropping the `*`, you shouldn't need that

Tried, still same result.

**Reply by B4nan** [reply]
Ideally make it a script, you shouldn't need web servers for this, it only makes things complex - reproductions should be minimal

**Reply by premdasvm** [reply]
@B4nan I made a reproduction repo https://github.com/premdasvm/Mikro-orm-custom-pivot-table-reproduction.git

On this reproduction i'm not getting the wrong query error tho, but the relation is not working like i expected it to be, most probably i've got the relation mapping messed up. 😅

**Reply by B4nan** [reply]
You dont clear the context in your code, your entities are already there and fully loaded. That `findOne` call you have there won't make any database query, as you search by the PK and the target is already there, including its populated relation.

**Reply by B4nan** [reply]
Btw doing this is also wrong:

```
entities: ["*.entity.js"],
entitiesTs: [User, Shift, UserShift],
```

If you use entity references, there is no need for folder based discovery, so this is the same as:

```
entities: [User, Shift, UserShift],
```

If you wanted to go with folder based discovery, it should be something like:

```
entities: ["*.entity.js"],
entitiesTs: ["*.entity.ts"],
```

**Reply by premdasvm** [reply]
> You dont clear the context in your code, your entities are already there and fully loaded. That `findOne` call you have there won't make any database query, as you search by the PK and the target is already there, including its populated relation.

clearing the context seemed to have my issue here. still facing the wrong query issue on my `NestJS` project tho. Now i'm sure that it's not an issue with the relation implementation. lemme look through a lil bit more. Thanks for the help.

**Reply by premdasvm** [reply]
@B4nan I was able to recreate my issue here.

What i did was i commented out the loading strategy from my `mikro-orm.config`, then the wrong query issue came back.

also when i had the `loadStrategy` as `JOINED` the `findOne` query being executed looked like this
```sql
select `u0`.*, `s1`.`id` as `s1__id`, `s1`.`name` as `s1__name` 
  from `user` as `u0` 
    left join `user_shift` as `u2` on `u0`.`id` = `u2`.`user_id` 
    left join `user` as `s1` on `u2`.`user_id` = `s1`.`id` 
where `u0`.`id` = 1
```

is it supposed to have that left join to `user` ?

**Reply by B4nan** [reply]
well, you want to select user's name, of course you need to join the user table

that query seems to be correct, it does not select anything else than PK and name of the relation, which is what you requested.

**Reply by premdasvm** [reply]
@B4nan 

Understood, Any idea what i'm doing wrong here to get the wrong query error?
```sql
`s0`.`id`, `s0`.`name` from `shift` as `s0` where `UserShift`.`shift` in (1)
```

I commented this out and the error seems to come back.
```
// loadStrategy: LoadStrategy.JOINED,
```

In my original project, the relations are defined the exact same way, but the wrong query executed there is different.

```sql
 select "s0"."id", "s0"."name" from "shift" as "s0" where "UserShift"."user" in (1)
 ```

**Reply by B4nan** [reply]
my guess is this is caused by your pivot entity extending your base entity - which adds another PK - that itself is not supported at all, composite key cannot have database defaults (this is limitation of the current implementation, otherwise it is valid on schema level)

**Reply by premdasvm** [reply]
@B4nan 

Wow. That was it. Finally. I removed `BaseEntity` from my pivot table and now it's working fine. Thanks @B4nan 

look like i've to modify my `BaseRepository` too i had custom `BaseRepository`

```ts
export class BaseRepository<T extends BaseEntity> extends EntityRepository<T> {
	/**
	 *
	 * @param {T} entity
	 * @return {*} {EntityManager}
	 * @memberof BaseRepository
	 */
	softRemove(entity: T): EntityManager {
		entity.deletedAt = new Date();
		this.persist(entity);

		return this.em;
	}
}
```
Any other recommend way to implement `softRemove` ?

**Reply by B4nan** [reply]
I don't have docs around it (will have a section in the new v6 guide), but you can use onFlush event subscriber to do this transparently, turning delete queries to update ones.

https://github.com/mikro-orm/mikro-orm/issues/1492#issuecomment-785394397

**Reply by B4nan** [reply]
Also note, that you don't need to call `em.persist()` on managed entities (loaded from db), not even for those created via `em.create()`.
